rclcpp/include/rclcpp/loaned_message.hpp

```diff
@@ -0,0 +1,170 @@
+// limitations under the License.
+
+#ifndef RCLCPP__LOANED_MESSAGE_HPP_
+#define RCLCPP__LOANED_MESSAGE_HPP_
+
+#include <memory>
+#include <utility>
+
+#include "rclcpp/logging.hpp"
+#include "rclcpp/publisher_base.hpp"
+
+#include "rcl/allocator.h"
+#include "rcl/publisher.h"
+
+namespace rclcpp
+{
+
+template<typename MessageT, typename AllocatorT = std::allocator<void>>
+class LoanedMessage
+{
+  using MessageAllocatorTraits = allocator::AllocRebind<MessageT, AllocatorT>;
+  using MessageAllocator = typename MessageAllocatorTraits::allocator_type;
+
+protected:
+  const rclcpp::PublisherBase * pub_;
+
+  std::unique_ptr<MessageT> message_;
+
+  const std::shared_ptr<MessageAllocator> message_allocator_;
+
+  /// Deleted copy constructor to preserve memory integrity
+  LoanedMessage(const LoanedMessage<MessageT> & other) = delete;
+
+public:
+  /// Constructor of the LoanedMessage class
+  /**
+   * The constructor of this class allocates memory for a given message type
+   * and associates this with a given publisher.
+   *
+   * \Note: Given the publisher instance, a case differentiation is being performaned
+   * which decides whether the underlying middleware is able to allocate the appropriate
+   * memory for this message type or not.
+   * In the case that the middleware can not loan messages, the passed in allocator instance
+   * is being used to allocate the message within the scope of this class.
+   * Otherwise, the allocator is being ignored and the allocation is solely performaned
+   * in the underlying middleware with its appropriate allocation strategy.
+   * The need for this arises as the user code can be written explicitly targeting a middleware
+   * capable of loaning messages.
+   * However, this user code is ought to be usable even when dynamically linked against
+   * a middleware which doesn't support message loaning in which case the allocator will be used.
+   *
+   * \param pub rclcpp::Publisher instance to which the memory belongs
+   * \param allocator Allocator instance in case middleware can not allocate messages
+   */
+  LoanedMessage(
+    const rclcpp::PublisherBase * pub,
+    const std::shared_ptr<std::allocator<MessageT>> allocator)
+  : pub_(pub),
+    message_(nullptr),
+    message_allocator_(allocator)
+  {
+    if (!pub) {
+      throw std::runtime_error("publisher pointer is null");
+    }
+
+    void * message_memory = nullptr;
+    if (pub_->can_loan_messages()) {
+      message_memory =
+        rcl_allocate_loaned_message(pub_->get_publisher_handle(), nullptr, sizeof(MessageT));
+    } else {
+      RCLCPP_WARN(
+        rclcpp::get_logger("rclcpp"),
+        "Currently used middleware can't loan messages. Local allocator will be used.");
+      message_memory = message_allocator_->allocate(1);
+    }
+    if (!message_memory) {
+      throw std::runtime_error("unable to allocate memory for loaned message");
+    }
+    message_.reset(new (message_memory) MessageT());
```

Comment:
I understand that it might have been called in the second case. I don't think it has been called in the first case though. The first one really only allocates enough space.

Reply:
I'm not sure I agree... What if it is reusing a message that was already there? Like reusing items from a pool? Why would the middleware not construct it after allocating?